inference-engine/src/mkldnn_plugin/nodes/mkldnn_roi_align_node.cpp

```diff
@@ -0,0 +1,363 @@
+            }
+        }
+        auto pool = [&] (int xBinInd_, int yBinInd_, int binOffsetInput_, int binOffsetOutput_, int blockResidual_) {
+            float pooledValue = 0;
+            unsigned int sampleIndex = 4 * (yBinInd_ * pooledW + xBinInd_) * numSamplesInBin;
+            for (unsigned int binSampleInd = 0; binSampleInd < numSamplesInBin; binSampleInd++) {
+                size_t part1Index = binOffsetInput_ + pointVector[sampleIndex].first * hInputStride +
+                                    pointVector[sampleIndex].second * wInputStride + blockResidual_;
+                float part1 = srcData[part1Index];
+                size_t part2Index = binOffsetInput_ + pointVector[sampleIndex + 1].first * hInputStride +
+                                    pointVector[sampleIndex + 1].second * wInputStride + blockResidual_;
+                float part2 = srcData[part2Index];
+                size_t part3Index = binOffsetInput_ + pointVector[sampleIndex + 2].first * hInputStride +
+                                    pointVector[sampleIndex + 2].second * wInputStride + blockResidual_;
+                float part3 = srcData[part3Index];
+                size_t part4Index = binOffsetInput_ + pointVector[sampleIndex + 3].first * hInputStride +
+                                    pointVector[sampleIndex + 3].second * wInputStride + blockResidual_;
+                float part4 = srcData[part4Index];
+
+                switch (opType) {
+                    case ROIAlignOpType::Max:
+                    {
+                        float sampleValue = std::max(
+                                {weightVector[sampleIndex] * part1,
+                                 weightVector[sampleIndex + 1] * part2,
+                                 weightVector[sampleIndex + 2] * part3,
+                                 weightVector[sampleIndex + 3] * part4});
+                        pooledValue = sampleValue > pooledValue ? sampleValue : pooledValue;
+                        break;
+                    }
+                    case ROIAlignOpType::Avg:
+                    default:
+                    {
+                        float sampleValue =
+                                weightVector[sampleIndex] * part1 +
+                                weightVector[sampleIndex + 1] * part2 +
+                                weightVector[sampleIndex + 2] * part3 +
+                                weightVector[sampleIndex + 3] * part4;
+                        pooledValue += sampleValue / numSamplesInBin;
+                    }
+                }
+                sampleIndex += 4;
+            }
+            size_t dstIndex = binOffsetOutput_ + yBinInd_ * hOutputStride +
+                              xBinInd_ * wOutputStride + blockResidual_;
+            dst[dstIndex] = pooledValue;
+        };
+        if (selectedFmt == mkldnn_nhwc) {
+            parallel_for(pooledH, [&](int yBinInd) {
+                parallel_for(pooledW, [&](int xBinInd) {
```

Comment:
From performance perspective it is better to use parallel_for2d instead of nested parallel loops.

Reply:
Done. Also added 3d parallel for for else-case